File: Dec2020/10Dec/BinaryTreePaths.cpp
/*
Problem name: 257. Binary Tree Paths
Problem link: https://leetcode.com/problems/binary-tree-paths/
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> ans;
    vector<string> binaryTreePaths(TreeNode* root) {
        if(!root)
            return ans;
        
        dfs(root, "", ans);
        return ans;
    }
    void dfs(TreeNode* root, string val, vector<string> &ans){
        val+=to_string(root->val);
        if(root->left == NULL && root->right == NULL){
            ans.push_back(val);
            return;
        }
        
        if(root->left)
            dfs(root->left, val+"->", ans);
        
        if(root->right)
            dfs(root->right, val+"->", ans);
        return;
    }
};

File: Dec2020/10Dec/BinaryTreePaths.java
/*
Problem Name: 257. Binary Tree Paths
Problem Link: https://leetcode.com/problems/binary-tree-paths/
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<String> ans;
    public List<String> binaryTreePaths(TreeNode root) {
        ans = new ArrayList<String>();
        traverse(root,"");
        return ans;
    }
    void traverse(TreeNode root,String path){
        if(root == null){
            return;
        }
        if(root.left == null && root.right == null){
            if(path.length() > 2)
                ans.add(path.substring(2)+"->"+root.val);
            else
                ans.add(""+root.val);
        }
        traverse(root.left,path+"->"+root.val);
        traverse(root.right,path+"->"+root.val);
    }
}
